src/fswAlgorithms/pointCloudProcessing/cameraTriangulation/cameraTriangulation.cpp

```diff
@@ -0,0 +1,228 @@
+    } else {
+        bskLogger.bskLog(BSK_ERROR, "cameraTriangulation: time tags from pointCloudInMsg and "
+                                    "keyPointsInMsg (timeTag_secondImage) are different, but should be equal.");
+        this->valid = false;
+    }
+    if (pointCloudSize != numberKeyPoints) {
+        bskLogger.bskLog(BSK_ERROR, "cameraTriangulation: number of features from pointCloudInMsg "
+                                    "(pointCloudSize) and keyPointsInMsg (numberKeyPoints) are different, "
+                                    "but should be equal.");
+        this->valid = false;
+    }
+}
+
+/*! This method writes the output messages each call of updateState
+    @param currentSimNanos current simulation time in nano-seconds
+    @return void
+*/
+void CameraTriangulation::writeMessages(uint64_t currentSimNanos)
+{
+    CameraLocalizationMsgPayload cameraLocationOutMsgBuffer;
+    cameraLocationOutMsgBuffer = this->cameraLocationOutMsg.zeroMsgPayload;
+
+    cameraLocationOutMsgBuffer.valid = this->valid;
+    cameraLocationOutMsgBuffer.cameraID = this->cameraID;
+    cameraLocationOutMsgBuffer.timeTag = this->timeTag;
+    Eigen::Vector3d sig_BN = eigenMRPd2Vector3d(this->sigma_BN);
+    eigenVector3d2CArray(sig_BN, cameraLocationOutMsgBuffer.sigma_BN);
+    eigenVector3d2CArray(this->estimatedCameraLocation, cameraLocationOutMsgBuffer.cameraPos_N);
+
+    this->cameraLocationOutMsg.write(&cameraLocationOutMsgBuffer, this->moduleID, currentSimNanos);
+}
+
+/*! This method performs the triangulation to estimate the unknown location given the known locations and images
+    @param knownLocations known positions (point cloud points)
+    @param imagePoints location of key points in pixel space
+    @param cameraCalibrationInverse inverse of camera calibration matrix
+    @param dcmCamera dcm from frame of interest F to camera frame C
+    @return Eigen::Vector3d
+*/
+Eigen::Vector3d CameraTriangulation::triangulation(std::vector<Eigen::Vector3d> knownLocations,
+                                                   std::vector<Eigen::Vector2d> imagePoints,
+                                                   std::vector<Eigen::Matrix3d> cameraCalibrationInverse,
+                                                   std::vector<Eigen::Matrix3d> dcmCamera)
+{
+    Eigen::Matrix3d Kinv = cameraCalibrationInverse.at(0); // inverse of camera calibration matrix
+    Eigen::Matrix3d dcm_CF = dcmCamera.at(0); // dcm from frame of interest F to camera frame C
+
+    unsigned long numPoints = knownLocations.size();
+
+    Eigen::MatrixXd A(3*numPoints, 3);
+    Eigen::VectorXd y(3*numPoints);
+
+    for (int c = 0; c < numPoints; ++c) {
+        // update camera calibration matrix and dcm in case they are different for each image
+        if (cameraCalibrationInverse.size() != 1) {
+            Kinv = cameraCalibrationInverse.at(c);
+        }
+        if (dcmCamera.size() != 1) {
+            dcm_CF = dcmCamera.at(c);
+        }
```

Comment:
Wouldn't these break if the cameraCalibrationInverse and dcmCamera vectors are not the same length (shorter) than the numPoints? It might be more general a problem than needs to be solved.

Reply:
Good point. I changed the function to take a constant Eigen Matrix3d as input for cameraCalibrationInverse and dcmCamera instead of a vector of matrices.